proposals/records/4-python-refactoring.md

```diff
@@ -0,0 +1,306 @@
+to the database.
+
+However, the `KnowledgeBox` class actually has responsibility for reading/writing various
+protobuf objects related to a KnowledgeBox, getting resources, creating shards and purging data.
+It is a mix of instance and class methods.
+
+
+#### Protobuf usage
+
+Protobuf being a type system and state (de)serialization system may make you think it is
+reasonable to use to communicate types internally; however, protobuf, especially for how
+we use it, is primarily a database schema definition system.
+
+When we use it to communicate state internally between components, we are exposing internal
+database storage details across our application. This is a data mapping pattern that often
+becomes difficult to manage.
+
+For example, imagine you have a relational database where data is stored in rows of fields.
+Then, you communicate database state as a dictionary across your application, directly
+mapping field names -> dictionary keys. What happens during a migration? What happens if
+you need to change the underlying implementation of the storage layer for a particular interface?
+You end up needing to update logic and/or state all over the code base. However, if you
+isolated those contracts to a data layer, your exposure to these sorts of problems is limited.
+
+### Global state
+
+A common pattern we use is to configure global components that are potentially
+used throughout the code base.
+
+Examples:
+
+- `get_storage`
+- `get_driver`
+- `get_cache`
+- `get_pubsub`
+- `get_partitioning`
+- `settings` -- all settings right now are initialized when a component is loaded
+- ...
+
+These components might be conditionally configured depending on what a top level component
+needs.
+
+This type of approach ends up encouraging continued increase in global state and injecting
+that global state into various components.
+
+Functions no longer do simple input/output since their input/output is a side effect of
+the global state.
+
+Finally, testing can be quite difficult because you always need to setup/teardown that global
+state OR monkey patch it.
+
+## Proposed Solution
+
+The proposal focusses on moving toward 3 patterns:
+
+1. Do not allow imports between our main components
+2. Isolate our data layers into separate interfaces
+3. Do not use global application state
+
+
+### Do not allow imports between our main components
+
+- `ingest` focuses on components for ingesting data and not providing interfaces to other components
+- shared tooling and data layers are moved to the `nucliadb.common` module
+- no main component should import from the `injest` module
+- nucliadb should stop using the `ingest` grpc service
+
+### Isolate our data layers into separate interfaces
+
+- create data layer management components
+- limit direct exposure of k/v and blob storage drivers
+- limit exposure of protobuf types(this might be very difficult and cumbersome)
```

Comment:
How are you planning to limit exposure with protobuf types?

Reply:
This is the part of the proposal I feel least good about.

I feel like our use of pb is wrong but an alternative would be mapping into internal data types which is a lot of churn and maybe not worth promoting too much at this point. 

An example of this can be found here: https://github.com/nuclia/nucliadb/blob/main/nucliadb/nucliadb/migrator/datamanager.py

the pb is only for storage but is not exposed

Reply:
Oh, I see, that's really nice. Ideally we have it like this (and also on the Rust side, don't propagate protos down the stack of services).

In the current status of things, might be unfeasible though... As we talked some times, our public API is also bound to our internal protos, so I think it'd need a global rethink on what we want to expose, whether an API with highly nested structures is worth...

Maybe we can start the refactor "ignoring" this, and at some point work on it